services/src/main/java/org/adamalang/services/social/Discord.java

```diff
@@ -22,46 +22,227 @@
+import java.util.TreeMap;
 import java.util.function.Consumer;
 
 public class Discord extends SimpleService {
   private static final Logger LOGGER = LoggerFactory.getLogger(Discord.class);
   private final FirstPartyMetrics metrics;
   private final WebClientBase base;
+  private final String botToken;
 
-  public Discord(FirstPartyMetrics metrics, WebClientBase base) {
+  private final String userToken;
+
+
+  public Discord(FirstPartyMetrics metrics, WebClientBase base, String botToken, String userToken) {
     super("discord", new NtPrincipal("discord", "service"), true);
     this.metrics = metrics;
     this.base = base;
+    this.botToken = botToken;
+    this.userToken = userToken;
   }
 
   public static Discord build(FirstPartyMetrics metrics, ServiceConfig config, WebClientBase base) throws ErrorCodeException {
-    return new Discord(metrics, base);
+    return new Discord(metrics, base, config.getString("bot_token", ""), config.getString("user_token", ""));
   }
 
+  class DiscordHttpResponder implements SimpleHttpResponder {
+    private final Logger logger;
+    private final RequestResponseMonitor.RequestResponseMonitorInstance monitor;
+    private final Callback<String> callback;
+    private ByteArrayOutputStream memory;
+    private boolean invokeSuccess;
+    private boolean emissionPossible;
+    private boolean logBody;
+
+    public DiscordHttpResponder(Logger logger, RequestResponseMonitor.RequestResponseMonitorInstance monitor, Callback<String> callback) {
+      this.logger = logger;
+      this.monitor = monitor;
+      this.callback = callback;
+      this.invokeSuccess = false;
+      this.emissionPossible = true;
+      this.logBody = false;
+    }
+    @Override
+    public void start(SimpleHttpResponseHeader header) {
+      if (emissionPossible) {
+        if (200 <= header.status && header.status <= 204) {
+          invokeSuccess = true;
+        } else {
+          emissionPossible = false;
+          logBody = HttpError.convert(header, logger,  ErrorCodes.WEB_STRING_CALLBACK_NOT_200, monitor, callback);
+        }
+      }
+    }
+
+    @Override
+    public void bodyStart(long size) {
+      if (size > 0) {
+        this.memory = new ByteArrayOutputStream((int) size);
+      }
+    }
+
+    @Override
+    public void bodyFragment(byte[] chunk, int offset, int len) {
+      if (memory == null) { // unknown size due to an error
+        this.memory = new ByteArrayOutputStream();
+      }
+      memory.write(chunk, offset, len);
+    }
+
+    @Override
+    public void bodyEnd() {
+      if (invokeSuccess && emissionPossible) {
+        String callbackVal = "";
+        if (memory != null) {
+          callbackVal = new String(memory.toByteArray(), StandardCharsets.UTF_8);
+        }
+        callback.success(callbackVal);
+        monitor.success();
+      }
+      if (logBody) {
+        if (memory != null) {
+          System.out.println(new String(memory.toByteArray(), StandardCharsets.UTF_8));
```

Comment:
So, this should use a Logger system this will block the current thread on stdout.